luaui/Widgets/unit_factory_quota.lua

```diff
@@ -0,0 +1,253 @@
+
+function widget:GetInfo()
+    return {
+      name = "Factory Quotas",
+      desc = "Creates quotas of units that should be fulfilled(for example 5 Sheldons, 5 Sumos), will be queued before factory queue.",
+      author = "hihoman23",
+      date = "2024",
+      license = "GNU GPL, v2 or later",
+      layer = -1,
+      enabled = true
+    }
+end
+
+VFS.Include('luarules/configs/customcmds.h.lua')
+
+local maxBuildProg = 0.075 -- maximum build progress that gets replaced in a repeat queue
+local maxMetal = 500 -- maximum metal cost that gets replaced in a repeat queue(7.5% of a juggernaut is still over 2k metal)
+
+-- factID is unitID of the factory
+local quotas = {} -- {[factID] = {[unitDefID] = amount, ...}, ...}
+
+local builtUnits = {} -- {[factID] = {[unitID] = true, ...}, ...}
+local unitFacts = {} -- {[unitID] = factID, ...}
+
+local possibleFacts = {}
+local factoryDefIDs = {}
+local metalcosts = {}
+
+for unitDefID, uDef in pairs(UnitDefs) do
+    metalcosts[unitDefID] = uDef.metalCost
+    if uDef.isFactory then
+        factoryDefIDs[unitDefID] = true
+        for _, opt in pairs(uDef.buildOptions) do
+            possibleFacts[opt] = possibleFacts[opt] or {}
+            possibleFacts[opt][unitDefID] = true
+        end
+    end
+end
+
+----- Speeeed ------
+local myTeam = Spring.GetMyTeamID()
+local GiveOrderToUnit = Spring.GiveOrderToUnit
+local GetFactoryCommands = Spring.GetFactoryCommands
+local GetUnitDefID = Spring.GetUnitDefID
+local GetRealBuildQueue = Spring.GetRealBuildQueue
+local GetUnitCmdDescs = Spring.GetUnitCmdDescs
+local FindUnitCmdDesc = Spring.FindUnitCmdDesc
+-----
+
+
+
+
+--------- quota logic -------------
+---@param table table
+---@param f function|nil
+local function findMin(table, f)
+    f = f or function(i1, i2, _, _) return i1 < i2 end
+    local best
+    local bestKey
+    for k, v in pairs(table) do
+        if not best then
+            best = v
+            bestKey = k
+        elseif f(v, best, k, bestKey) then
+            best = v
+            bestKey = k
+        end
+    end
+    return best, bestKey
+end
+
+---length of all tables
+---@param table table
+---@return integer
+function table.length(table)
+    local x = 0
+    for _ in pairs(table or {}) do
```

Comment:
```suggestion
    for _ in pairs(table) do
```
Nil should not be treated as an empty table, this will swallow errors from passing nil.